### PR TITLE
speed up sub-feature in row-wise parallelism

### PR DESCRIPTION
**Pull request by guolinke**
- [ ] update related documents

### CODE REVIEW

File: src/io/bin.cpp [hunk 1]
```diff
@@ -666,21 +666,38 @@ namespace LightGBM {
...
     const double multi_val_bin_sparse_threshold = 0.25f;
     if (sparse_rate >= multi_val_bin_sparse_threshold) {
-      if (num_bin <= 256) {
-        return new MultiValSparseBin<uint8_t>(num_data, num_bin);
-      } else if (num_bin <= 65536) {
-        return new MultiValSparseBin<uint16_t>(num_data, num_bin);
-      } else {
-        return new MultiValSparseBin<uint32_t>(num_data, num_bin);
-      }
+      const double average_element_per_row = sparse_rate * num_feature;
```
Comment (guolinke) [inline src/io/bin.cpp]:
```suggestion
      const double average_element_per_row = (1.0 - sparse_rate) * num_feature;
```

File: src/io/dataset.cpp [hunk 2]
```diff
@@ -1099,51 +1130,195 @@ void Dataset::DumpTextFile(const char* text_filename) {
...
 
-void Dataset::ConstructHistogramsMultiVal(const MultiValBin* multi_val_bin, const data_size_t* data_indices, data_size_t num_data,
-                                          const score_t* gradients, const score_t* hessians,
-                                          bool is_constant_hessian,
-                                          hist_t* hist_data) const {
-  Common::FunctionTimer fun_time("Dataset::ConstructHistogramsMultiVal", global_timer);
+
+void Dataset::InitTrain(const std::vector<int8_t>& is_feature_used,
+                        bool is_colwise, TrainingTempState* temp_state) const {
+  Common::FunctionTimer fun_time("Dataset::InitTrain", global_timer);
```
Comment (guolinke) [inline src/io/dataset.cpp]:
```suggestion
  Common::FunctionTimer fun_time("Dataset::InitTrain", global_timer);
  temp_state->use_subfeature = false;
```

File: src/io/dataset.cpp [hunk 3]
```diff
@@ -1099,51 +1130,195 @@ void Dataset::DumpTextFile(const char* text_filename) {
...
+        sum_used_dense_ratio += dense_rate;
+      }
+      sum_dense_ratio += dense_rate;
+      ++total;
+    }
+  }
+  global_timer.Stop("Dataset::InitTrain.Prep");
+  const double k_subfeature_threshold = 0.6;
+  if (sum_used_dense_ratio >= sum_dense_ratio * k_subfeature_threshold) {
+    temp_state->use_subfeature = false;
```
Comment (guolinke) [inline src/io/dataset.cpp]:
```suggestion
```

File: src/io/dataset.cpp [hunk 4]
```diff
@@ -1099,51 +1130,195 @@ void Dataset::DumpTextFile(const char* text_filename) {
...
-  Common::FunctionTimer fun_time("Dataset::ConstructHistogramsMultiVal", global_timer);
+
+void Dataset::InitTrain(const std::vector<int8_t>& is_feature_used,
+                        bool is_colwise, TrainingTempState* temp_state) const {
+  Common::FunctionTimer fun_time("Dataset::InitTrain", global_timer);
+  temp_state->use_subfeature = false;
+  if (temp_state->multi_val_bin == nullptr) {
+    return;
+  }
+  global_timer.Start("Dataset::InitTrain.Prep");
```
Comment (StrikerRUS) [inline src/io/dataset.cpp]: Is it OK to have `Prep` twice?

File: src/io/dataset.cpp [hunk 5]
```diff
@@ -1099,51 +1130,195 @@ void Dataset::DumpTextFile(const char* text_filename) {
...
+      ++total;
+    }
+  }
+  global_timer.Stop("Dataset::InitTrain.Prep");
+  const double k_subfeature_threshold = 0.6;
+  if (sum_used_dense_ratio >= sum_dense_ratio * k_subfeature_threshold) {
+    return;
+  }
+  temp_state->use_subfeature = true;
+  global_timer.Start("Dataset::InitTrain.Prep");
```
Comment (StrikerRUS) [inline src/io/dataset.cpp]: Is it OK to have `Prep` twice?

File: src/treelearner/serial_tree_learner.cpp [hunk 6]
```diff
@@ -80,15 +80,18 @@ void SerialTreeLearner::Init(const Dataset* train_data, bool is_constant_hessian
...
   } else {
     // cannot change is_hist_col_wise during training
-    multi_val_bin_.reset(dataset->TestMultiThreadingMethod(ordered_gradients_.data(), ordered_hessians_.data(), is_feature_used_,
+    temp_state_.reset(dataset->TestMultiThreadingMethod(
+        ordered_gradients_.data(), ordered_hessians_.data(), is_feature_used_,
       is_constant_hessian_, is_hist_colwise_, !is_hist_colwise_, &is_hist_colwise_));
   }
 }
 
+// Todo: optimized bagging for multi-val bin
```
Comment (StrikerRUS) [inline src/treelearner/serial_tree_learner.cpp]: Create feature request for TODO?

File: src/io/dataset.cpp (hunk 5 above)
Comment (guolinke) [reply src/io/dataset.cpp]: Yes, to avoid the break by return.

Comment (StrikerRUS) [reply src/io/dataset.cpp]: Ah, I see. Thanks for the explanation!

**Comment by StrikerRUS** [comment]
@guolinke

File: src/treelearner/serial_tree_learner.cpp (hunk 6 above)
Comment (guolinke) [reply src/treelearner/serial_tree_learner.cpp]: yeah, will create one

**Pull request merged**